src/avx512-16bit-common.h

```diff
@@ -259,6 +259,38 @@ X86_SIMD_SORT_INLINE type_t get_pivot_16bit(type_t *arr,
     return ((type_t *)&sort)[16];
 }
 
+template <typename vtype, typename type_t>
+static void
+qsort_partial_16bit_(int64_t k, type_t *arr,
+                     int64_t left, int64_t right,
+                     int64_t max_iters)
+{
+    /*
+     * Resort to std::sort if quicksort isnt making any progress
+     */
+    if (max_iters <= 0) {
+        std::sort(arr + left, arr + right + 1, comparison_func<vtype>);
+        return;
+    }
+    /*
+     * Base case: use bitonic networks to sort arrays <= 128
+     */
+    if (right + 1 - left <= 128) {
+        sort_128_16bit<vtype>(arr + left, (int32_t)(right + 1 - left));
+        return;
+    }
+
+    type_t pivot = get_pivot_16bit<vtype>(arr, left, right);
+    type_t smallest = vtype::type_max();
+    type_t biggest = vtype::type_min();
+    int64_t pivot_index = partition_avx512<vtype>(
+            arr, left, right + 1, pivot, &smallest, &biggest);
+    if ((pivot != smallest) && (k <= pivot_index))
+        qsort_partial_16bit_<vtype>(k, arr, left, pivot_index - 1, max_iters - 1);
+    else if ((pivot != biggest) && (k > pivot_index))
```

Comment:
What happens when `pivot != smallest` and `pivot == biggest` and `k > pivot_index`? Won't you exit the function with [`arr[0], arr[k]`] not yet sorted?

Reply:
oh, wait. I'm a little confused. Is `avx512_qsort_partial<T>` equivalent to `std::partial_sort`? or is it just to extract the `k'th` largest element? Looking at the tests  in `test_qsort_partial.hpp`, it looks like you care about the k^th element only, in which case I think the name of the function is very misleading.

Reply:
`avx512_qsort_partial<T>` is supposed to put the `k^th` smallest element into position `k`, while the `avx512_qsort_partialrange<T>` is supposed to ensure that the elements between the `kfrom^th` smallest to the `kto^th` smallest are in their correct sorted position (where the k's are cardinal ranks starting from 1).

Good point about the naming, there is some discrepancy. The `_partialrange` function is similar to `std::partial_sort`, but differs in how to interpret the arguments (see above). `std::partial_sort` uses the offset from to determine the positions between which the elements should be in sorted order (the same as `qsort_64bit_`, `qsort_32bit_`, etc).

I just found `std::nth_element` which behaves similarly to how I've implemented `_partial`. I think it might be a good idea to change the code in order to better align with these two functions (also in using an offset as the argument).

Reply:
Following that change, I can add the benchmarks that compare with `std::partial_sort` and `std::nth_element`. What do you think?

I also wanted to thank you for taking the time to review my PR and provide your constructive input.

Reply:
Code has significantly changed so I'm marking this as resolved. Please consider the latest version of the branch.